File: web/src/add.ts
import { existsSync } from "node:fs";
import { readFile, stat } from "node:fs/promises";
import { join } from "node:path";

import { IgnoreParser } from "./IgnoreParser.ts";
import { status } from "./status.ts";
import { getIndex, hashObject, writeIndex } from "./utils.ts";

export async function add(path: string, files: string[]): Promise<void> {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		throw new Error("Not a git repository");
	}

	const indexPath = join(gitDir, "index");
	const index = await getIndex(indexPath);

	// Load gitignore patterns
	const gitignore = new IgnoreParser();
	await gitignore.loadGitignore(path);

	for (const file of files) {
		if (gitignore.isIgnored(file)) {
			continue;
		}

		const fullPath = join(path, file);

		if (!existsSync(fullPath)) {
			throw new Error(`File not found: ${file}`);
		}

		const content = await readFile(fullPath, "utf-8");
		// Write blob object to .git/objects and get hash
		const hash = await hashObject(gitDir, content, "blob");
		const stats = await stat(fullPath);

		index.set(file, {
			path: file,
			sha: hash,
			mode: "100644",
			size: stats.size,
			ctimeSeconds: Math.floor(stats.ctimeMs / 1000),
			ctimeNanos: (stats.ctimeMs % 1000) * 1000000,
			mtimeSeconds: Math.floor(stats.mtimeMs / 1000),
			mtimeNanos: (stats.mtimeMs % 1000) * 1000000,
			dev: stats.dev,
			ino: stats.ino,
			uid: stats.uid,
			gid: stats.gid,
		});
	}

	await writeIndex(indexPath, index);
}

export async function addAll(path: string): Promise<void> {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		throw new Error("Not a git repository");
	}

	const currentStatus = await status(path);
	const filesToAdd = [...currentStatus.untracked, ...currentStatus.modified];

	if (filesToAdd.length > 0) {
		await add(path, filesToAdd);
	}

	if (currentStatus.deleted.length > 0) {
		const indexPath = join(gitDir, "index");
		const index = await getIndex(indexPath);
		for (const file of currentStatus.deleted) {
			index.delete(file);
		}
		await writeIndex(indexPath, index);
	}
}

File: dotnet/Gitologist/Add.cs
using Gitologist.Types;

namespace Gitologist;

public static class Add
{
    public static async Task AddFiles(string path, string[] files)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        // Load gitignore patterns
        var gitignore = new IgnoreParser();
        await gitignore.LoadGitignore(path);

        var indexPath = Path.Combine(gitDir, "index");
        var index = await Utils.GetIndex(indexPath);

        foreach (var file in files)
        {
            var fullPath = Path.Combine(path, file);

            if (!File.Exists(fullPath))
            {
                throw new FileNotFoundException($"File not found: {file}");
            }

            if (gitignore.IsIgnored(file))
            {
                continue;
            }

            var content = await File.ReadAllTextAsync(fullPath);
            // Write blob object to .git/objects and get hash
            var hash = await Utils.HashObject(gitDir, content, "blob");
            var fileInfo = new FileInfo(fullPath);

            var ctimeSeconds = (uint)fileInfo.CreationTime.ToUniversalTime().Subtract(DateTime.UnixEpoch).TotalSeconds;
            var ctimeNanos = (uint)((fileInfo.CreationTime.ToUniversalTime().Subtract(DateTime.UnixEpoch).TotalSeconds % 1) * 1_000_000_000);
            var mtimeSeconds = (uint)fileInfo.LastWriteTime.ToUniversalTime().Subtract(DateTime.UnixEpoch).TotalSeconds;
            var mtimeNanos = (uint)((fileInfo.LastWriteTime.ToUniversalTime().Subtract(DateTime.UnixEpoch).TotalSeconds % 1) * 1_000_000_000);

            var normalizedFile = file.Replace('\\', '/');
            index[normalizedFile] = new IndexEntry
            {
                Path = normalizedFile,
                Sha = hash,
                Mode = "100644",
                Size = (uint)fileInfo.Length,
                CtimeSeconds = ctimeSeconds,
                CtimeNanos = ctimeNanos,
                MtimeSeconds = mtimeSeconds,
                MtimeNanos = mtimeNanos,
                Dev = 0,
                Ino = 0,
                Uid = 0,
                Gid = 0
            };
        }

        await Utils.WriteIndex(indexPath, index);
    }

    public static async Task AddAll(string path)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        var currentStatus = await Status.GetStatus(path);
        var filesToAdd = new List<string>();
        filesToAdd.AddRange(currentStatus.Untracked);
        filesToAdd.AddRange(currentStatus.Modified);

        if (filesToAdd.Count > 0)
        {
            await AddFiles(path, filesToAdd.ToArray());
        }

        if (currentStatus.Deleted.Length > 0)
        {
            var indexPath = Path.Combine(gitDir, "index");
            var index = await Utils.GetIndex(indexPath);
            foreach (var file in currentStatus.Deleted)
            {
                index.Remove(file);
            }
            await Utils.WriteIndex(indexPath, index);
        }
    }
}
